README.md
<h2>Version: 0.2.1</h2>

<h1>Usage:</h1>

<h3>Password</h3>

```ruby

Nretnil::Password.generate
# => {:password=>"6SoiSo979F1IUva", :phonetic=>"6 SIERRA oscar india SIERRA oscar 9 7 9 FOXTROT 1 INDIA UNIFORM victor alpha"}

Nretnil::Password.generate(25,true)
# => {:password=>"m0#~~2fC)xQ9FrA9EhV2ya7WZ", :phonetic=>"mike 0 (hash) (tilda) (tilda) 2 foxtrot CHARLIE (right-parenthesis) x-ray QUEBEC 9 FOXTROT romeo ALPHA 9 ECHO hotel VICTOR 2 yankee alpha 7 WHISKEY ZULU"}

Nretnil::Password.generate(20,false)
# => {:password=>"7eS92RRJYIQFK33L9hhP", :phonetic=>"7 echo SIERRA 9 2 ROMEO ROMEO JULIET YANKEE INDIA QUEBEC FOXTROT KILO 3 3 LIMA 9 hotel hotel PAPA"}

Nretnil::Password.generate(11)
# => {:password=>"sw1czI13iTt", :phonetic=>"sierra whiskey 1 charlie zulu INDIA 1 3 india TANGO tango"}

```

<h3>Passphrase</h3>

```ruby

Nretnil::Password.phrase
# => {:password=>"earthyellowmuskdeersitbike", :phonetic=>"earth yellow musk deer sit bike", :sym_sub_pass=>"3@rthy3ll0wmu$kd33r$!tb!k3", :sym_sub_phon=>"3@rth y3ll0w mu$k d33r $!t b!k3"}

```

<h3>Custom</h3>

```ruby

Nretnil::Password.custom(20, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 'a', 'b', 'c', 'd', 'e', 'f'])
# => {:password=>"4aed41ac2bf4846ae3e0"}

Nretnil::Password.custom(15, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 'a', 'b', 'c', 'd', 'e', 'f'])[:password]
# => 6a94a8a763b4c3b

```

<h3>Phonetic</h3>

```ruby

password="aBcD 1234 !@#$"

Nretnil::Password.to_phonetic(password)
# => (ALPHA) (CAPITAL-BRAVO) (CHARLIE) (CAPITAL-DELTA) (SPACE) (WUN) (TOO) (TREE) (ROW ER) (SPACE) (EXCLAMATION-POINT) (AT) (HASH) (DOLLAR-SIGN)

```

<h3>UUID</h3>

```ruby

Nretnil::Password.uuid
# => c48ff793-aacb-e4de-2cb1-ef525633c42f

```

<h3>Symbol Substitution</h3>

```ruby

Nretnil::Password.sym_sub("string")
# => $tr!ng

```


## Testing
```
bundle Install
bundle exec rspec spec/spec_netnil_password.rb
```


CHANGELOG.md
<h4>0.2.1</h4>

* Removed trailing space from phonetic version

<h4>0.2.0</h4>

* Cleaned up code with rubocop and added unit tests

<h4>0.1.7</h4>

* to_phonetic(password) function

<h4>0.1.6</h4>

* Updated Words and Colors to use same set as FakeData

<h4>0.1.5</h4>

* sym_sub(string) function

<h4>0.1.4</h4>

* Rubocop Updates

<h4>0.1.3</h4>

* Updated phonetic output for password symbols to be (symbol)
* Set password default length to 15

<h4>0.1.2</h4>

* Updated @@ to Static Variables as they should have been
* Updated Password to class from Module so it can be instansiated.

<h4>0.1.1</h4>

* uuid generator fix

<h4>0.1.0</h4>

* Intial Release

<h4>0.0.4</h4>

* rename manual to custom

<h4>0.0.3</h4>

* Fix for generate function

<h4>0.0.2</h4>

* Added uuid feature

<h4>0.0.1</h4>

* Built Initial Gem
* Three features
* Password Generator
* Passphrase Generator
* Password Generator with specified characters


spec/spec_helper.rb
#!/bin/env ruby
# frozen_string_literal: true

require 'bundler/setup'
Bundler.setup

APP_ROOT = File.expand_path(File.join(File.dirname(__FILE__), '..'))
$LOAD_PATH << File.join(APP_ROOT, 'lib/nretnil') # so rspec knows where your file could be

require 'nretnil-password'

RSpec.configure do |config|
  # some (optional) config here
end


spec/spec_nretnil_password.rb
# frozen_string_literal: true

require 'spec_helper'

describe 'character_with_symbols' do
  before do
    @pass = Nretnil::Password.new
  end

  it 'should produce a hash with a char and photnec equivalent' do
    expect(@pass.character_with_symbols.is_a?(Hash)).to be true
    expect(@pass.character_with_symbols[:a].nil?).to be false
    expect(@pass.character_with_symbols[:p].nil?).to be false
  end
end

describe 'character_no_symbols' do
  before do
    @pass = Nretnil::Password.new
    @character_set = @pass.character_no_symbols
  end

  it 'should produce a hash with a char and photnec equivalent' do
    expect(@character_set.is_a?(Hash)).to be true
    expect(@character_set[:a].nil?).to be false
    expect(@character_set[:p].nil?).to be false
    expect(@pass.symbols.include?(@character_set[:a])).to be false
    expect(@pass.phonetic_symbols.include?(@character_set[:p])).to be false
  end
end

describe 'get_next' do
  before do
    @pass = Nretnil::Password.new
  end

  it 'should produce a hash with a char and photnec equivalent' do
    expect(@pass.get_next.is_a?(Hash)).to be true
    expect(@pass.get_next[:a].nil?).to be false
    expect(@pass.get_next[:p].nil?).to be false
  end
end

describe 'to_phonetic' do
  before do
    @password = '0n3T20Thr33'
  end

  it '("0n3T20Thr33") should produce a string with correct phonetic parts' do
    expect(Nretnil::Password.to_phonetic(@password)).to eq '(ZE-RO) (NOVEMBER) (TREE) (CAPITAL-TANGO) (TOO) (ZE-RO) (CAPITAL-TANGO) (HOTEL) (ROMEO) (TREE) (TREE)'
    expect(Nretnil::Password.to_phonetic(@password).split(' ').length).to eq @password.length
  end
end

describe 'Nretnil::Password.generate' do
  before do
  end

  it 'should produce 15 char password' do
    expect(Nretnil::Password.generate.is_a?(Hash)).to be true
    expect(Nretnil::Password.generate[:password].length).to be 15
    expect(Nretnil::Password.generate[:phonetic].split(' ').length).to be 15
  end

  it 'generate(25) should produce 25 char password' do
    expect(Nretnil::Password.generate(25).is_a?(Hash)).to be true
    expect(Nretnil::Password.generate(25)[:password].length).to be 25
    expect(Nretnil::Password.generate(25)[:phonetic].split(' ').length).to be 25
  end

  it 'generate(25, true) should produce 25 char password' do
    expect(Nretnil::Password.generate(25, true).is_a?(Hash)).to be true
    expect(Nretnil::Password.generate(25, true)[:password].length).to be 25
    expect(Nretnil::Password.generate(25, true)[:phonetic].split(' ').length).to be 25
  end
end

describe 'Nretnil::Password.custom' do
  before do
    @custom_chars = %w[a b c 1 2 3]
  end

  it '(10, %w[a b c 1 2 3]) should produce a 10 char password' do
    password = Nretnil::Password.custom(10, @custom_chars)
    expect(password[:password].length).to be 10
    password[:password].split('').each do |char|
      expect(@custom_chars.include?(char)).to be true
    end
    expect(Nretnil::Password.custom(10, @custom_chars).is_a?(Hash)).to be true
  end
end

describe 'Nretnil::Password.uuid' do
  before do
    @uuid_regex = /[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}/
  end

  it 'should produce a 36 char uuid' do
    expect(Nretnil::Password.uuid.length).to be 36
  end

  it 'should be a string' do
    expect(Nretnil::Password.uuid.is_a?(String)).to be true
  end

  it 'should produce a valid uuid' do
    expect(Nretnil::Password.uuid =~ @uuid_regex).not_to be nil
  end
end

describe 'Nretnil::Password.phrase' do
  before do
  end

  it 'should produce a string' do
    expect(Nretnil::Password.phrase.is_a?(Hash)).to be true
    expect(Nretnil::Password.phrase[:phonetic].is_a?(String)).to be true
  end

  it 'should have at least 4 words' do
    expect(Nretnil::Password.phrase.is_a?(Hash)).to be true
    expect(Nretnil::Password.phrase[:phonetic].split(' ').length).to be > 3
  end
end

describe 'Nretnil::Password.sym_sub' do
  before do
    @string = 'TestString'
  end

  it 'should produce a string with equal length' do
    expect(Nretnil::Password.sym_sub(@string).length == @string.length).to be true
  end

  it 'should produce a string that properly substitutes' do
    expect(Nretnil::Password.sym_sub(@string)).to eq 'T3$t$tr!ng'
  end
end


lib/nretnil/password/data.rb
#!/bin/env ruby
# frozen_string_literal: true

require 'rubygems'

module Nretnil
  # Password Class
  class Password
    SYMBOLS = ['~', '!', '@', '#', '$', '%', '^', '&', '*', '(', ')', '-', '+'].freeze
    PHONETIC_SYMBOLS = ['(tilda)', '(exclamation-point)', '(at)', '(hash)', '(dollar-sign)', '(percent-sign)', '(carrot)', '(ampersand)', '(star)', '(left-parenthesis)', '(right-parenthesis)', '(dash)', '(plus)'].freeze
    ALPHA = %w[a b c d e f g h i j k l m n o p q r s t u v w x y z].freeze
    PHONETIC_ALPHA = ['alpha', 'bravo', 'charlie', 'delta', 'echo', 'foxtrot', 'golf', 'hotel', 'india', 'juliet', 'kilo', 'lima', 'mike', 'november', 'oscar', 'papa', 'quebec', 'romeo', 'sierra', 'tango', 'uniform', 'victor', 'whiskey', 'x-ray', 'yankee', 'zulu'].freeze
    PHONETIC_NUMBERS = ['ze-ro', 'wun', 'too', 'tree', 'row er', 'fife', 'six', 'seven', 'ait', 'niner'].freeze
    HEX = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 'a', 'b', 'c', 'd', 'e', 'f'].freeze

    def symbols
      SYMBOLS
    end

    def phonetic_symbols
      PHONETIC_SYMBOLS
    end

    def alpha
      ALPHA
    end

    def phonetic_alpha
      PHONETIC_ALPHA
    end

    def phonetic_numbers
      PHONETIC_NUMBERS
    end

    def hex
      HEX
    end
  end
end


lib/nretnil/password/password.rb
#!/bin/env ruby
# frozen_string_literal: true

require 'rubygems'

module Nretnil
  # Password Class
  class Password
    SUBSTITUTIONS = { a: '@', e: '3', i: '!', o: '0', s: '$' }.freeze

    def character_with_symbols
      pass = Password.new
      type_rand = rand(10)
      if type_rand < 3 # number
        a = p = rand(10)
      elsif type_rand > 7 # symbol
        sym_rand = rand(pass.symbols.length)
        a = pass.symbols[sym_rand]
        p = pass.phonetic_symbols[sym_rand]
      else # letter
        case_rand = rand(2)
        sym_rand = rand(pass.alpha.length)
        a = case_rand.zero? ? pass.alpha[sym_rand] : pass.alpha[sym_rand].upcase
        p = case_rand.zero? ? pass.phonetic_alpha[sym_rand] : pass.phonetic_alpha[sym_rand].upcase
      end
      { a: a, p: p }
    end

    def character_no_symbols
      pass = Password.new
      type_rand = rand(10)
      if type_rand < 4 # number
        a = p = rand(10)
      else # letter
        case_rand = rand(2)
        sym_rand = rand(pass.alpha.length)
        a = case_rand.zero? ? pass.alpha[sym_rand] : pass.alpha[sym_rand].upcase
        p = case_rand.zero? ? pass.phonetic_alpha[sym_rand] : pass.phonetic_alpha[sym_rand].upcase
      end
      { a: a, p: p }
    end

    def get_next(symbols = false)
      pass = Password.new
      return pass.character_with_symbols if symbols

      pass.character_no_symbols
    end

    def self.to_phonetic(password)
      return false if password.nil?

      pass = Password.new
      new_pass = ''
      password.split('').each do |c|
        new_char = ''
        cap = (c == c.upcase)
        space = c.strip.empty?
        alpha = pass.alpha.find_index(c.downcase)
        sym = pass.symbols.find_index(c)
        new_char = '(' + pass.phonetic_alpha[alpha] + ')' unless alpha.nil?
        new_char = pass.phonetic_symbols[sym].upcase unless sym.nil?
        new_char = '(' + pass.phonetic_numbers[c.to_i] + ')' if alpha.nil? && sym.nil? && !space
        new_char = '(SPACE)' if space
        new_pass += cap && !alpha.nil? ? new_char.upcase.gsub('(', '(CAPITAL-') : new_char.upcase
        new_pass += ' '
      end
      new_pass.strip
    end

    def self.generate(length = 15, symbols = false)
      pass = Password.new
      password = ''
      phonetic = ''
      (1..length).each do |_i|
        char = pass.get_next(symbols)
        password += char[:a].to_s
        phonetic += char[:p].to_s + ' '
      end

      { password: password, phonetic: phonetic.strip }
    end

    def self.custom(length, characters = [])
      password = ''

      (1..length).each do |_i|
        password += characters[rand(characters.length)].to_s
      end

      { password: password }
    end

    def self.uuid
      pass = Password.new
      uuid = ''
      (0...32).each do |i|
        uuid += '-' if [8, 12, 16, 20].include? i
        uuid += pass.hex[rand(16)].to_s
      end
      uuid
    end

    def self.phrase
      pass = Password.new
      phonetic = pass.colors[rand(pass.colors.length)][:name].downcase + ' ' + pass.animals[rand(pass.animals.length)] + ' ' + pass.verbs[rand(pass.verbs.length)] + ' ' + pass.nouns[rand(pass.nouns.length)]
      password = phonetic.delete(' ')
      { password: password, phonetic: phonetic, sym_sub_pass: sym_sub(password), sym_sub_phon: sym_sub(phonetic) }
    end

    def self.sym_sub(string)
      new_string = ''
      string.split(//).each do |c|
        new_string += SUBSTITUTIONS[c.downcase.to_sym].nil? ? c : SUBSTITUTIONS[c.downcase.to_sym]
      end
      new_string
    end
  end
end


examples/example.rb
#!/bin/env ruby
# frozen_string_literal: true

require 'rubygems'
require 'nretnil-password'

puts '<h3>Password</h3>'
puts "\n"

puts '```ruby'
puts "\n"

puts 'Nretnil::Password.generate'
puts '# => ' + Nretnil::Password.generate.to_s
puts "\n"

puts 'Nretnil::Password.generate(25,true)'
puts '# => ' + Nretnil::Password.generate(25, true).to_s
puts "\n"

puts 'Nretnil::Password.generate(20,false)'
puts '# => ' + Nretnil::Password.generate(20, false).to_s
puts "\n"

puts 'Nretnil::Password.generate(11)'
puts '# => ' + Nretnil::Password.generate(11).to_s
puts "\n"

puts '```'
puts "\n"

puts '<h3>Passphrase</h3>'
puts "\n"

puts '```ruby'
puts "\n"

puts 'Nretnil::Password.phrase'
puts '# => ' + Nretnil::Password.phrase.to_s
puts "\n"

puts '```'
puts "\n"

puts '<h3>Custom</h3>'
puts "\n"

puts '```ruby'
puts "\n"

puts 'Nretnil::Password.custom(20, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, \'a\', \'b\', \'c\', \'d\', \'e\', \'f\'])'
puts '# => ' + Nretnil::Password.custom(20, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 'a', 'b', 'c', 'd', 'e', 'f']).to_s
puts "\n"

puts 'Nretnil::Password.custom(15, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, \'a\', \'b\', \'c\', \'d\', \'e\', \'f\'])[:password]'
puts '# => ' + Nretnil::Password.custom(15, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 'a', 'b', 'c', 'd', 'e', 'f'])[:password]
puts "\n"

puts '```'
puts "\n"

puts '<h3>Phonetic</h3>'
puts "\n"

puts '```ruby'
puts "\n"

password = 'aBcD 1234 !@#$'

puts "password=\"#{password}\"\n\n"

puts 'Nretnil::Password.to_phonetic(password)'
puts '# => ' + Nretnil::Password.to_phonetic(password).to_s
puts "\n"

puts '```'
puts "\n"

puts '<h3>UUID</h3>'
puts "\n"

puts '```ruby'
puts "\n"

puts 'Nretnil::Password.uuid'
puts '# => ' + Nretnil::Password.uuid
puts "\n"

puts '```'
puts "\n"

puts '<h3>Symbol Substitution</h3>'
puts "\n"

puts '```ruby'
puts "\n"

puts 'Nretnil::Password.sym_sub("string")'
puts '# => ' + Nretnil::Password.sym_sub('string')
puts "\n"

puts '```'
puts "\n"
